File: lib_elevator_control/src/main/java/com/reeman/agv/elevator/request/ApiClient.java
package com.reeman.agv.elevator.request;


import com.reeman.commons.constants.Constants;
import com.reeman.agv.elevator.BuildConfig;
import com.reeman.agv.elevator.request.service.ApiService;
import com.reeman.commons.utils.AESUtil;
import com.reeman.commons.utils.MMKVManager;

import java.io.IOException;
import java.security.SecureRandom;
import java.security.cert.X509Certificate;
import java.util.concurrent.TimeUnit;

import javax.net.ssl.SSLContext;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;
import okio.Buffer;
import okio.BufferedSource;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava3.RxJava3CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import timber.log.Timber;

public class ApiClient {

    private final String TAG = BuildConfig.ELEVATOR_DIR;
    private final ApiService apiService;
    private static volatile ApiClient instance;

    public ApiClient() {
        OkHttpClient.Builder httpClientBuilder = new OkHttpClient.Builder();
        httpClientBuilder
                .connectTimeout(15, TimeUnit.SECONDS)
                .writeTimeout(15,TimeUnit.SECONDS)
                .addInterceptor(new RetryInterceptor(1, 2))
                .addInterceptor(new CustomHeaderInterceptor())
                .addInterceptor(new LoggingInterceptor());
        try {
            SSLContext sslContext = SSLContext.getInstance("TLS");
            sslContext.init(null, new TrustManager[]{new X509TrustManager() {
                @Override
                public void checkClientTrusted(X509Certificate[] chain, String authType) {
                }

                @Override
                public void checkServerTrusted(X509Certificate[] chain, String authType) {
                }

                @Override
                public X509Certificate[] getAcceptedIssuers() {
                    return new X509Certificate[0];
                }
            }}, new SecureRandom());
            httpClientBuilder.sslSocketFactory(sslContext.getSocketFactory());
            httpClientBuilder.hostnameVerifier((hostname, session) -> true); // 忽略主机名验证
        } catch (Exception e) {
            e.printStackTrace();
        }
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(BuildConfig.ELEVATOR_URL)
//                .baseUrl("")
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava3CallAdapterFactory.create())
                .client(httpClientBuilder.build())
                .build();

        apiService = retrofit.create(ApiService.class);
    }

    public static ApiClient getInstance() {
        if (instance == null) {
            synchronized (ApiClient.class) {
                if (instance == null) {
                    instance = new ApiClient();
                }
            }
        }
        return instance;
    }

    public ApiService getApiService() {
        return apiService;
    }


    private static class CustomHeaderInterceptor implements Interceptor {
        @Override
        public Response intercept(Chain chain) throws IOException {
            Request originalRequest = chain.request();
            Request.Builder requestBuilder = originalRequest.newBuilder();
            requestBuilder.header("SECURITY-KEY", Constants.DEFAULT_SECURITY_KEY);
            requestBuilder.header("TENANT_ID", Constants.DEFAULT_TENANT_ID);
            Request newRequest = requestBuilder.build();
            return chain.proceed(newRequest);
        }
    }

    public class LoggingInterceptor implements Interceptor {

        @Override
        public Response intercept(Chain chain) throws IOException {
            Request request = chain.request();

            // 打印请求信息
            Timber.tag(TAG).w("Request URL: %s", request.url());
            Timber.tag(TAG).w("Request Method: %s", request.method());
            Timber.tag(TAG).v("Request Headers: %s", request.headers());

            // 打印请求体信息
            if (request.body() != null) {
                Buffer buffer = new Buffer();
                request.body().writeTo(buffer);
                Timber.tag(TAG).w("Request Body: %s", buffer.readUtf8());
            }

            // 执行请求
            Response response = chain.proceed(request);

            // 打印响应信息
            Timber.tag(TAG).w("Response Code: %s", response.code());
            Timber.tag(TAG).v("Response Headers: %s", response.headers());

            // 打印响应体信息
            ResponseBody responseBody = response.body();
            if (responseBody != null) {
                BufferedSource source = responseBody.source();
                source.request(Long.MAX_VALUE); // 读取整个响应体
                Buffer buffer = source.buffer();
                Timber.tag(TAG).w("Response Body: %s", buffer.clone().readUtf8());
            }

            return response;
        }
    }

    private class RetryInterceptor implements Interceptor {

        private final int maxRetries;
        private final int retryDelaySeconds;

        public RetryInterceptor(int maxRetries, int retryDelaySeconds) {
            this.maxRetries = maxRetries;
            this.retryDelaySeconds = retryDelaySeconds;
        }

        @Override
        public Response intercept(Chain chain) throws IOException {
            Request request = chain.request();
            Response response = null;
            IOException exception = null;
            for (int retry = 0; retry < maxRetries; retry++) {
                try {
                    response = chain.proceed(request);
                    if (response.isSuccessful()) {
                        break;
                    }
                } catch (IOException e) {
                    exception = e;
                    try {
                        Timber.tag(TAG).w("请求失败,网络错误,自动重试中...");
                        Thread.sleep(retry * retryDelaySeconds * 1000L);
                    } catch (InterruptedException interruptedException) {
                        Thread.currentThread().interrupt();
                    }
                }


            }

            if (response == null && exception != null) {
                throw exception;
            }

            return response;
        }
    }
}

File: AGV_v1.9.0/lib_dispatch/src/main/java/com/reeman/dispatch/request/ApiClient.kt
package com.reeman.dispatch.request

import com.reeman.commons.utils.SpManager
import com.reeman.dispatch.BuildConfig
import com.reeman.dispatch.constants.Constants
import com.reeman.dispatch.model.request.RoomLoginReq
import com.reeman.dispatch.request.service.ApiService
import kotlinx.coroutines.runBlocking
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Response
import okio.Buffer
import retrofit2.Retrofit
import retrofit2.adapter.rxjava3.RxJava3CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import timber.log.Timber
import java.io.IOException
import java.security.SecureRandom
import java.security.cert.X509Certificate
import java.util.concurrent.TimeUnit
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManagerFactory
import javax.net.ssl.X509TrustManager

object ApiClient{

    private const val TAG = BuildConfig.DISPATCH_DIR
    var token: String? = null
    var roomName: String? = null
    var roomPwd: String? = null
    var host:String?=null
    var hostname:String?=null

    private val httpClient: OkHttpClient by lazy {
        buildHttpClient()
    }

    private fun buildHttpClient(): OkHttpClient {
        val builder = OkHttpClient.Builder()
            .connectTimeout(5, TimeUnit.SECONDS)
            .writeTimeout(5, TimeUnit.SECONDS)
            .readTimeout(10, TimeUnit.SECONDS)
            .addInterceptor(CustomTokenInterceptor())
            .addInterceptor(LoggingInterceptor())
            .addInterceptor(RetryInterceptor(listOf("/task/create"),20,2000L))

        try {
            val sslContext = SSLContext.getInstance("TLS").apply {
                init(null, arrayOf(object : X509TrustManager {
                    override fun checkClientTrusted(chain: Array<X509Certificate>, authType: String) {}
                    override fun checkServerTrusted(chain: Array<X509Certificate>, authType: String) {}
                    override fun getAcceptedIssuers(): Array<X509Certificate> = arrayOf()
                }), SecureRandom())
            }
            builder.sslSocketFactory(sslContext.socketFactory, TrustManagerFactory.getInstance("X509").trustManagers[0] as X509TrustManager)
            builder.hostnameVerifier { _, _ -> true }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return builder.build()
    }

    fun getApiService(): ApiService {
        val retrofit = Retrofit.Builder()
            .baseUrl("http://127.0.0.1")
            .addConverterFactory(GsonConverterFactory.create())
            .addCallAdapterFactory(RxJava3CallAdapterFactory.create())
            .client(httpClient)
            .build()

        return retrofit.create(ApiService::class.java)
    }

    private class CustomTokenInterceptor : Interceptor {
        override fun intercept(chain: Interceptor.Chain): Response {
            var request = chain.request()
            val sp = SpManager.getInstance()

            if (!request.url().toString().endsWith("login")) {
                if (token.isNullOrEmpty()) {
                    token = sp.getString(Constants.DISPATCH_TOKEN, "")
                }
                request = request.newBuilder()
                    .header("Authorization", "Bearer $token")
                    .build()
            }

            var response = chain.proceed(request)
            if (response.code() == 401 && !request.url().toString().endsWith("login") && !host.isNullOrBlank() && !hostname.isNullOrBlank()) {
                if (!roomName.isNullOrEmpty() && !roomPwd.isNullOrEmpty()) {
                    Timber.tag(TAG).w("Token expired: $token")
                    runBlocking {
                        val mResponse = getApiService().loginRoomSync(Urls.getLoginUrl(host!!), RoomLoginReq(roomName!!, roomPwd!!, hostname!!, 1))
                        if (mResponse.isSuccessful){
                            mResponse.body()?.data?.get("token")?.let {
                                token = it
                                sp.edit().putString(Constants.DISPATCH_TOKEN, token).apply()
                                response.close()
                                request = request.newBuilder()
                                    .header("Authorization", "Bearer $token")
                                    .build()

                                response = chain.proceed(request)
                            }
                        }
                    }
                }
            }
            return response
        }
    }



    private class LoggingInterceptor : Interceptor {
        override fun intercept(chain: Interceptor.Chain): Response {
            val request = chain.request()

            Timber.tag(TAG).w("Request URL: ${request.url()}")
            Timber.tag(TAG).w("Request Method: ${request.method()}")
//            Timber.tag(TAG).v("Request Headers: ${request.headers()}")

            request.body()?.let {
                val buffer = Buffer()
                it.writeTo(buffer)
                Timber.tag(TAG).w("Request Body: ${buffer.readUtf8()}")
            }

            val response = chain.proceed(request)

            Timber.tag(TAG).w("Response Code: ${response.code()}")
            Timber.tag(TAG).v("Response Headers: ${response.headers()}")

            response.body()?.let { responseBody ->
                val source = responseBody.source()
                source.request(Long.MAX_VALUE)
                val buffer = source.buffer
                Timber.tag(TAG).w("Response Body: ${buffer.clone().readUtf8()}")
            }

            return response
        }
    }

    class RetryInterceptor(
        private val urlSuffixes: List<String>,
        private val maxRetryCount: Int = 3,
        private val retryDelayMillis: Long = 1000L
    ) : Interceptor {
        override fun intercept(chain: Interceptor.Chain): Response {
            var request = chain.request()
            var response: Response? = null
            var retryCount = 0

            val urlMatches = urlSuffixes.any { request.url().toString().endsWith(it) }
            if (!urlMatches)return chain.proceed(request)

            while (retryCount < maxRetryCount) {
                try {
                    response = chain.proceed(request)
                    if (response.isSuccessful || response.code() != 500) {
                        return response
                    }
                } catch (e: IOException) {
                    if (retryCount >= maxRetryCount - 1) {
                        Timber.tag(TAG).e("Request failed after $maxRetryCount retries: ${e.message}")
                        throw e
                    } else {
                        Timber.tag(TAG).w("Retrying... (${retryCount + 1}/$maxRetryCount)")
                    }
                }

                retryCount++
                Thread.sleep(retryDelayMillis)
            }

            return response ?: throw IOException("Failed to get response after $maxRetryCount retries.")
        }
    }

}
